crates/ide-completion/src/completions/extern_crate.rs

```diff
@@ -0,0 +1,96 @@
+//! Completion for extern crates
+
+use ide_db::{FxHashSet, SymbolKind};
+use syntax::{
+    ast::{self, NameRef},
+    AstNode,
+};
+
+use crate::{context::CompletionContext, CompletionItem, CompletionItemKind};
+
+use super::Completions;
+
+pub(crate) fn complete_extern_crate(
+    acc: &mut Completions,
+    ctx: &CompletionContext<'_>,
+    name_ref: Option<&NameRef>,
+) {
+    let imported_extern_crates: FxHashSet<String> = ctx
+        .token
+        .parent_ancestors()
+        .find_map(ast::SourceFile::cast)
+        .map(|src_file| {
+            src_file
+                .syntax()
+                .children()
+                .into_iter()
+                .filter_map(ast::ExternCrate::cast)
+                .filter_map(|node| node.name_ref())
+                .map(|name| name.to_string())
+                .collect()
+        })
+        .unwrap_or_default();
```

Comment:
Going via syntax is pretty messy and error prone (given macros could create these as well), but we can't go the HIR route here right now as we don't record `extern_crate` decls in the HIR yet. Can you put a FIXME above this something like `FIXME: Collect extern crate names via HIR when possible`?

Reply:
Actually with https://github.com/rust-lang/rust-analyzer/pull/15377/ we can make this work on the HIR

Reply:
Nice it felt like a hack haha. 

I made a quick implementation using the resolver to perform external crate decl queries [here](https://github.com/rust-lang/rust-analyzer/blob/5d3fd84449d48aee53d0f9c6dfc4e7164dcb4d1e/crates/hir-def/src/resolver.rs#L483C44-L483C44). Am I on the right track or should I be finding the information somewhere else? I suspect a bit of adjustment somewhere might remove the need for querying extern crate decl data :thinking:

Reply:
Yup, using the query is the correct approach